### PR TITLE
Unit Tests Jest

### PR DESCRIPTION
**Pull request by Daniel-PerezF**
Screen Shot 2023-11-09 at 8 53 10 PM

### CODE REVIEW

File: unit-tests-jest/numbers.test.ts [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+test('should return an array with even numbers', () => {
+  const result1 = evenNumbers([1, 2, 3, 4, 5, 6, 7, 8]);
+  expect(result1).toEqual([2, 4, 6, 8]);
+
+  const result2 = evenNumbers([2, 4, 6, 7]);
+  expect(result2).toEqual([2, 4, 6]);
+
+  const result3 = evenNumbers([0]);
+  expect(result3).toEqual([0]);
+});
```
Comment (RobertGardner) [inline unit-tests-jest/numbers.test.ts:12]: A good edge case to test is what happens if you pass an empty array.

File: unit-tests-jest/numbers.test.ts [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+test('should return a number formatted in dollars and cents', () => {
+  const result1 = toDollars(1);
+  expect(result1).toEqual('$1.00');
+
+  const result2 = toDollars(444);
+  expect(result2).toEqual('$444.00');
+});
```
Comment (RobertGardner) [inline unit-tests-jest/numbers.test.ts:20]: You should test that, for example, 3.45 results in '$3.45'.
Another one is that, for example, 3.4567 results in '$3.46'.

File: unit-tests-jest/numbers.test.ts [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+test('Returns a new array of numbers where every entry has been divided by the given divisor. Does not modify the original array', () => {
+  const result1 = divideBy([1, 2, 3], 1);
+  expect(result1).toEqual([1, 2, 3]);
+
+  const result2 = divideBy([2, 3, 4, 6], 2);
+  expect(result2).toEqual([1, 1.5, 2, 3]);
+
+  const result3 = divideBy([0.2, 3.49], 2);
+  expect(result3).toEqual([0.1, 1.745]);
+});
```
Comment (RobertGardner) [inline unit-tests-jest/numbers.test.ts:31]: A good edge case to test is what happens to an empty array. Or division by 0.

It would also be good to test that the original array is not modified, as the spec says.

File: unit-tests-jest/numbers.test.ts [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+test('Modifies an object by multiplying the value of each key, but only if that value is a number. Key names are arbitrary', () => {
+  const result1 = multiplyBy({ one: 1, two: '2' }, 1);
+  expect(result1).toEqual({ one: 1, two: '2' });
+
+  const result2 = multiplyBy({ one: '1', two: '2', three: 3 }, 3);
+  expect(result2).toEqual({ one: '1', two: '2', three: 9 });
+  const result3 = multiplyBy({ one: null, two: undefined }, 2);
+  expect(result3).toEqual({ one: null, two: undefined });
+});
```
Comment (RobertGardner) [inline unit-tests-jest/numbers.test.ts:41]: A good edge case is what happens with an empty object.

File: unit-tests-jest/numbers.test.ts [hunk 5]
```diff
@@ -0,0 +1,41 @@
+import { evenNumbers, toDollars, divideBy, multiplyBy } from './numbers';
+
+test('should return an array with even numbers', () => {
```
Comment (RobertGardner) [inline unit-tests-jest/numbers.test.ts:3]: The best way to lay out your tests is to use `describe`, which each test case a separate test inside the describe.

For example:
```
describe('evenNumbers', () => {
  test('returns even numbers', () => {
    const result1 = evenNumbers([1, 2, 3, 4, 5, 6, 7, 8]);
    expect(result1).toEqual([2, 4, 6, 8]);
  });

  test('works correctly with empty array', () => {
    const result1 = evenNumbers([]);
    expect(result1).toEqual([]);
  });
});
```

File: unit-tests-jest/numbers.ts [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+ */
+export function evenNumbers(numbers: number[]): number[] {
+  return numbers.filter((n) => n % 2 === 0);
+}
+
+/**
+ * Returns a number formatted in dollars and cents.
+ */
+export function toDollars(amount: number): string {
+  return `$${amount}.00`;
```
Comment (RobertGardner) [inline unit-tests-jest/numbers.ts:12]: This is not a correct implementation. It doesn't work for non-integer numbers. Take a look at String.toFixed.

File: unit-tests-jest/numbers.ts [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+): Record<string, unknown> {
+  const result: Record<string, unknown> = {};
+  Object.entries(obj).forEach(([key, value]) => {
+    if (typeof value === 'number') {
+      result[key] = Number(value) * multiplier;
+    } else {
+      result[key] = value;
+    }
+  });
+  return result;
```
Comment (RobertGardner) [inline unit-tests-jest/numbers.ts:43]: The spec says it "Modifies an object" but this implementation creates a new object. It should be changed to modify the object. It can be trivially changed to work correctly by changing lines 38 and 40 to use `obj[key]` instead of `result[key]`.

**Review by RobertGardner — APPROVED** [review]
Some things that could improved at your leisure.
